src/nbl/video/CVKLogicalDevice.h

```diff
@@ -178,52 +802,382 @@ class CVKLogicalDevice final : public ILogicalDevice
     {
-        return nullptr;
+        // Todo(achal): Hoist creation out of constructor
+        return nullptr; // return core::make_smart_refctd_ptr<CVulkanFramebuffer>(this, std::move(params));
     }
 
-    core::smart_refctd_ptr<IGPUSpecializedShader> createGPUSpecializedShader_impl(const IGPUShader* _unspecialized, const asset::ISpecializedShader::SInfo& _specInfo, const asset::ISPIRVOptimizer* _spvopt) override
+    // Todo(achal): For some reason this is not printing shader errors to console
+    core::smart_refctd_ptr<IGPUSpecializedShader> createGPUSpecializedShader_impl(const IGPUShader* _unspecialized, const asset::ISpecializedShader::SInfo& specInfo, const asset::ISPIRVOptimizer* spvopt) override
     {
-        return nullptr;
+        if (_unspecialized->getAPIType() != EAT_VULKAN)
+            return nullptr;
+
+        const CVulkanShader* unspecializedShader = static_cast<const CVulkanShader*>(_unspecialized);
+
+        const std::string& entryPoint = specInfo.entryPoint;
+        const asset::ISpecializedShader::E_SHADER_STAGE shaderStage = specInfo.shaderStage;
+
+        core::smart_refctd_ptr<asset::ICPUBuffer> spirv = nullptr;
+        if (unspecializedShader->containsGLSL())
+        {
+            const char* begin = reinterpret_cast<const char*>(unspecializedShader->getSPVorGLSL()->getPointer());
+            const char* end = begin + unspecializedShader->getSPVorGLSL()->getSize();
+            std::string glsl(begin, end);
+            core::smart_refctd_ptr<asset::ICPUShader> glslShader_woIncludes =
+                m_physicalDevice->getGLSLCompiler()->resolveIncludeDirectives(glsl.c_str(),
+                    shaderStage, specInfo.m_filePathHint.string().c_str());
+
+            spirv = m_physicalDevice->getGLSLCompiler()->compileSPIRVFromGLSL(
+                reinterpret_cast<const char*>(glslShader_woIncludes->getSPVorGLSL()->getPointer()),
+                shaderStage, entryPoint.c_str(), specInfo.m_filePathHint.string().c_str());
+        }
+        else
+        {
+            spirv = unspecializedShader->getSPVorGLSL_refctd();
+        }
+
+        // Should just do this check in ISPIRVOptimizer::optimize
+        if (!spirv)
+            return nullptr;
+
+        if (spvopt)
+            spirv = spvopt->optimize(spirv.get(), m_physicalDevice->getDebugCallback()->getLogger());
+
+        if (!spirv)
+            return nullptr;
+
+        VkShaderModuleCreateInfo vk_createInfo = { VK_STRUCTURE_TYPE_SHADER_MODULE_CREATE_INFO };
+        vk_createInfo.pNext = nullptr;
+        vk_createInfo.flags = static_cast<VkShaderModuleCreateFlags>(0); // reserved for future use by Vulkan
+        vk_createInfo.codeSize = spirv->getSize();
+        vk_createInfo.pCode = reinterpret_cast<const uint32_t*>(spirv->getPointer());
+
+        VkShaderModule vk_shaderModule;
+        if (vkCreateShaderModule(m_vkdev, &vk_createInfo, nullptr, &vk_shaderModule) == VK_SUCCESS)
```

Comment:
create the shader module in `createGPUShader`

Reply:
the only reason why OpenGL backend keeps raw GLSL up to this point is because it needed to know the all the shaders in the pipline for the compilation

Reply:
Resolved

Reply:
Still a TODO: <https://github.com/Devsh-Graphics-Programming/Nabla/blob/erfan_merge/src/nbl/video/CVulkanShader.h#L30>

@Erfan-Ahmadi Vulkan backend should not be holding onto the spirv buffer after shader module creation, shouldn't even go into the `CVulkanShader` constructor!


Also https://github.com/Devsh-Graphics-Programming/Nabla/blob/erfan_merge/src/nbl/video/CVulkanLogicalDevice.h#L635
this clone is pointless if the real shader contents are GLSL, some shaders are a few megs of GLSL, useless memory pressure.